src/main.rs

```diff
@@ -1,245 +1,202 @@
-
-    // FIXME write Wasm::Value -> Ipld converter
-    let boxed_results: Box<[Value]> = _function
-        .call(&mut store, wasm_args.as_slice())
-        .expect("tried to call function");
-
-    let res: String = boxed_results
-        .into_vec()
-        .iter()
-        .map(|v| v.to_string())
-        .collect();
-
-    let resource = Url::parse(format!("ipfs://{}", wasm).as_str()).expect("IPFS URL");
-
-    let ipld_args: Ipld = Ipld::List(
-        wasm_args
-            .iter()
-            .map(|a| match a {
-                Value::I32(i) => Ipld::from(*i),
-                _ => todo!(),
-            })
-            .collect(),
-    );
-
-    let closure = workflow::closure::Closure {
-        resource,
-        action: workflow::closure::Action::from("wasm/run"),
-        inputs: workflow::closure::Input::IpldData(ipld_args),
-    };
+            println!("Fuck: {providers:?}");
 
-    let closure_ipld: Ipld = closure.clone().into();
-
-    let mut closure_bytes = Vec::new();
-    closure_ipld
-        .encode(DagCborCodec, &mut closure_bytes)
-        .expect("CBOR Serialization");
-
-    let closure_cid2: String = <Closure as Into<Link<Closure>>>::into(closure.clone())
-        .to_string_of_base(Base::Base32HexLower)
-        .expect("string CID");
-
-    let mut conn = establish_connection();
-
-    let new_receipt = NewReceipt {
-        val: res.parse::<i32>().expect("i32"), // FIXME!
-        closure_cid: closure_cid2,             // WHYYYYYYY
-    };
+            let requests = providers.into_iter().map(|p| {
+                let name = name.clone();
+                let mut client = client.clone();
+                async move { client.request_file(p, name).await }.boxed()
+            });
 
-    diesel::insert_into(receipts::table)
-        .values(&new_receipt)
-        .execute(&mut conn)
-        .expect("Error saving new post");
+            let file_content = future::select_ok(requests)
+                .await
+                .map_err(|_| anyhow!("None of the providers returned file."))?
+                .0;
 
-    todo!("advertise receipt");
+            io::stdout().write_all(&file_content)?
 
-    let res_copy = res.clone().into_bytes();
-
-    println!("Wasm CID: {}", closure_cid2);
-    println!("Result value: {}", res.to_string());
-
-    let res_cursor = Cursor::new(res);
-    let output = match ipfs.add(res_cursor).await.expect("a CID") {
-        AddResponse { hash, .. } => hash,
-    };
-
-    println!("Result CID: {}", output);
-
-    network_client.start_providing(closure_cid2).await;
-
-    loop {
-        match network_events.next().await {
-            Some(network::Event::InboundRequest { request, channel }) => {
-                if request == closure_cid2 {
-                    network_client.respond_file(res_copy.clone(), channel).await;
-                }
-            }
-            e => todo!("{:?}", e),
+            //
+        }
+        Argument::Provide { wasm, fun, args } => {
```

Comment:
This feels like a big block to just put inside a branch, no? Is there a cost to breaking this up and/or why not move it to another module?

Reply:
So, I'll put a todo here. I don't want the big block, but I was rethinking ways to abstract this out. So, I put it back in one spot before coming back to it.